feat: schedule implementation (#369)
defenseunicorns/pepr: docs/OnSchedule.md

Reviewer: I feel like the doc is missing one-or-more notes about how OnSchedule'll be have if the "(optional)" inputs aren't set (i.e. every, unit, startTime, completions).  Not sure if that's better put as an overall note, or as something noted next to each field , like a default-value-if-not-set  note or something.  :thinking:

Author: i hear ya, i was hoping the intellisense would cover that but i can make it explicit

Author: sorry, misinterpreted, so you would like to me to add what happens if the optional values are not set. I can cover that. I think more documentation is generally better.

# OnSchedule

The `OnSchedule` feature allows you to schedule and automate the execution of specific code at predefined intervals or schedules. This feature is designed to simplify recurring tasks and can serve as an alternative to traditional CronJobs. This code is designed to be run at the top level on a Capability, not within a function like `When`. 
  

> > **Note -** To use this feature in dev node `npx pepr dev`, you MUST set `PEPR_WATCH_MODE="true"`. This is because the scheduler only runs in the watch controller not started in dev mode by default. `PEPR_WATCH_MODE="true" npx pepr dev`.

Reviewer: ```suggestion
> **Note -** To use this feature in dev mode you MUST set `PEPR_WATCH_MODE="true"`. This is because the scheduler only runs on the watch controller and the watch controller is not started by default in dev mode.

For example: `PEPR_WATCH_MODE="true" npx pepr dev`
```

Typo / flow.

## Best Practices

> `OnSchedule` is designed for targeting intervals equal to or larger than 30 seconds due to the storage mechanism used to archive schedule info, anything less than that is _use at your own risk_.

Reviewer: ```suggestion
`OnSchedule` is designed for targeting intervals equal to or larger than 30 seconds due to the storage mechanism used to archive schedule info.
```

Feels like a stronger statement if we just tell people that they shouldn't use <30 sec intervals -- don't want to suggest that people could use shorter intervals _even_ at their own risk, IMO, if we don't want to support that. :thinking:

  
## Usage

> Create a schedule by calling the OnSchedule function with the following parameters:

Reviewer: ```suggestion
Create a recurring task execution by calling the OnSchedule function with the following parameters:
```

Swapped subject to put a focus on the "thing to be accomplished" (i.e. the scripted task) rather than on the mechanism for accomplishing it (i.e. the schedule).

**name** - The unique name of the schedule.

> **every** - (Optional) An integer that represents the frequency of the schedule.

Reviewer: ```suggestion
**every** - (Optional) An integer that represents the frequency of the schedule in number of _units_.
```

My first read of this made me wonder what unit of measure was being used... which is, of course, clarified below, but this edit should help others avoid the same (brief) confusion.

**unit** - (Optional) A string specifying the time unit for the schedule (e.g., `seconds`, `minute`, `minutes`, `hour`, `hours`).  

**startTime** - (Optional) A UTC timestamp indicating when the schedule should start. All date times must be provided in GMT.  

**run** - A function that contains the code you want to execute on the defined schedule.  

> **completions** - (Optional) An integer indicating the maximum number of times the schedule should run.

Reviewer: ```suggestion
**completions** - (Optional) An integer indicating the maximum number of times the schedule should run to completion.
```

Minor edit to put focus on how OnSchedule isn't just _starting_ some number of times, it's _completing_ a certain number of runs -- helps clarify why the field is named as it is, IMO.

## Examples

> Update the curr ConfigMap every 15 seconds and use the store to track the current count:

Reviewer: ```suggestion
Update a ConfigMap every 30 seconds:
```

Cut "curr" since it's a code item & not really representative of the ConfigMap itself.

Update time period to match 1) >30 sec limitation (described above), and 2) the example code (laid out below).

Cut note about "the store" since it's an implementation detail / not really important for the thrust of the example.

```typescript

> let curr = 0 

Reviewer: What is this variable for, actually?

Author: not needed, meant to take it out. will remove

OnSchedule({
    name: "hello-interval",
    every: 30,
    unit: "seconds",
    run: async () => {
      Log.info("Wait 30 seconds and create/update a ConfigMap");

      try {
        await K8s(kind.ConfigMap).Apply({
          metadata: {
            name: "last-updated",
            namespace: "default",
          },
          data: {
            count: `${new Date()}`,
          },
        });

      } catch (error) {
        Log.error(error, "Failed to apply ConfigMap using server-side apply.");
      }
    },
  });
```

> Every 24 hours refresh the AWSToken, delays the start for 30 seconds, and only run 3 times:

Reviewer: ```suggestion
Refresh an AWSToken every 24 hours, with a delayed start of 30 seconds, running a total of 3 times:
```

Reorg for flow.

```typescript

OnSchedule({
  name: "refresh-aws-token",
  every: 24,
  unit: "hours",
  startTime: new Date(new Date().getTime() + 1000 * 30),
  run: async () => {
    await RefreshAWSToken();
  },
  completions: 3,
});
```

## Advantages 

- Simplifies scheduling recurring tasks without the need for complex CronJob configurations.
- Provides flexibility to define schedules in a human-readable format.
- Allows you to execute code with precision at specified intervals.
- Supports limiting the number of schedule completions for finite tasks.

## Considerations

- Ensure that all date times are provided in GMT to maintain consistent scheduling.

> - Review your scheduling needs to determine if OnSchedule is suitable for your use case compared to traditional CronJobs.

Reviewer: I'm a little unclear on the intended function of this line -- what is it warning me about?  Is there any sort of statement we could make about why someone would choose to look into using a CronJob over using OnSchedule that might imply the suitability of one over the other?

If not, this line might not be needed..? :thinking:

Author: I think it is more of an ADR type doc, due to that I removed it.

> - OnSchedule waits for the store to be ready before executing because the schedule is stored in etcd. This means that if the store is not ready, the schedule will not run.

Reviewer: ```suggestion
- OnSchedule is backed by the Pepr Store so it has to wait for the store to be ready before executing.  This means that there will be a (slight) delay between the time Pepr starts up and the time OnSchedule can begin executing scheduled tasks.
```

Etcd is an implementation detail of the Pepr Store & not all that clarifying... unless you're trying to get the reader to look into the details of the Pepr Store (which doesn't really seem to be the case), so I cut it out.
